pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go

```diff
@@ -0,0 +1,465 @@
+		return
+	}
+
+	klog.V(2).Infof("Waiting for application of %s label", discovery.LabelSkipMirror)
+	c.waitForSkipLabelApplication()
+	klog.V(2).Infof("%s label successfully applied", discovery.LabelSkipMirror)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
+	}
+
+	<-stopCh
+}
+
+// This controller should not run until the EndpointSlice skip mirroring label
+// has been applied to Endpoints managed by the Endpoints controller. This
+// function waits until that's true.
+func (c *Controller) waitForSkipLabelApplication() error {
+	return wait.PollImmediate(2*time.Second, 5*time.Minute, func() (bool, error) {
+		endpoints, err := c.endpointsLister.List(labels.Everything())
+		if err != nil {
+			return false, err
+		}
+		services, err := c.serviceLister.List(labels.Everything())
+		if err != nil {
+			return false, err
+		}
+
+		servicesByName := map[string]*v1.Service{}
+		for _, svc := range services {
+			servicesByName[svc.Name] = svc
+		}
+
+		numEndpointsWithProblems := 0
+		for _, ep := range endpoints {
+			svc, ok := servicesByName[ep.Name]
+			if ok && svc.Spec.Selector != nil {
+				labelVal, ok := ep.Labels[discovery.LabelSkipMirror]
+				if !ok || labelVal != "true" {
+					klog.V(4).Infof("Expected %s Endpoints resource to have %s label set to true, got %s", ep.Name, discovery.LabelSkipMirror, labelVal)
+					numEndpointsWithProblems++
+				}
+			}
+		}
+
+		if numEndpointsWithProblems > 0 {
+			klog.V(2).Infof("Waiting to start EndpointSliceMirroring controller until Endpoints controller applies %s label (waiting for %d/%d Endpoints)", discovery.LabelSkipMirror, numEndpointsWithProblems, len(endpoints))
+			return false, nil
+		}
+
+		return true, nil
+	})
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same service
+// at the same time
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncEndpoints(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		klog.Warningf("Error mirroring EndpointSlices for %q Endpoints, retrying. Error: %v", key, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Retry budget exceeded, dropping %q Endpoints out of the queue: %v", key, err)
+	c.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (c *Controller) syncEndpoints(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing EndpointSlices for %q Endpoints. (%v)", key, time.Since(startTime))
+	}()
+
```

Comment:
I've added this but I'm a little concerned over the potential noise since it's mostly duplicating the log message above. In practice the log message above is already relatively spammy in kube-controller-manager. On the other hand, if/when things really break it's helpful to know that we at least reached the syncEndpoints() function.

Reply:
What is the standard log level of the controller? V(4) expectation is log spam..

Reply:
That's fair. OSS clusters default to v(4) but that's likely rarely used for prod clusters.